File: other-programming-languages/R/2024/2024-001/part-002/main.r
# Define the file path
file_path <- "input.txt"

# Read the file into a data frame
df <- read.table(file_path, header = FALSE)

# Extract the two columns into two lists
left_list <- df[[1]]
right_list <- df[[2]]

# Convert the right list into a table (dictionary equivalent in R)
right_list_counts <- table(right_list)

# Calculate the similarity score
similarity_score <- sum(sapply(left_list, function(left) {
  count <- ifelse(left %in% names(right_list_counts), right_list_counts[as.character(left)], 0)
  left * count
}))

# Print the similarity score
print(paste("Similarity Score:", similarity_score))

File: other-programming-languages/Kotlin/2024/2024-001/part-002/Main.kt
import java.io.File

fun main() {
    // Define the file path
    val filePath = "input.txt"

    // Read the file into two lists
    val (leftList, rightList) = readInputFile(filePath)

    // Convert the right list into a frequency map
    val rightListCounts = rightList.groupingBy { it }.eachCount()

    // Calculate the similarity score
    val similarityScore = leftList.sumOf { left -> left * (rightListCounts[left] ?: 0) }

    // Print the similarity score
    println("Similarity Score: $similarityScore")
}

// Function to read the input file and extract two lists
fun readInputFile(filePath: String): Pair<List<Int>, List<Int>> {
    val leftList = mutableListOf<Int>()
    val rightList = mutableListOf<Int>()

    File(filePath).useLines { lines ->
        lines.forEach { line ->
            val parts = line.split("\\s+".toRegex())
            if (parts.size == 2) {
                leftList.add(parts[0].toInt())
                rightList.add(parts[1].toInt())
            }
        }
    }

    return Pair(leftList, rightList)
}
